Validate routing commands using updated routing state

vigyasharma opened the pull request:
When multiple commands are called in sequence, fetch shards
from mutable, up-to-date routing nodes to ensure each command's
changes are visible to subsequent commands.

This addresses an issue uncovered during work on #41050.

vigyasharma committed:
Fix allocation commands to use routing nodes to fetch shards

When multiple commands are called in sequence, fetching shards
from mutable, upto date routing nodes ensures each command's
changes are visible to subsequent commands.

elasticmachine commented:
Pinging @elastic/es-distributed

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java:
```diff
@@ -110,13 +110,22 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
         }
 
-        final ShardRouting shardRouting;
+        // Validate input shard and index values
```
I think the behaviour of this block of code is clear without a comment.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -101,20 +102,36 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
         }
 
-        final ShardRouting primaryShardRouting;
+        // Validate input shard and index values
```
I think the behaviour of this block of code is clear without a comment.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java:
```diff
@@ -108,13 +108,22 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
         }
 
-        final ShardRouting shardRouting;
+        // Validate input shard and index values
```
I think the behaviour of this block of code is clear without a comment.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java:
```diff
@@ -110,13 +110,22 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
+
+        ShardRouting shardRouting = null;
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        for (ShardRouting shard: unassigned) {
```
Nit: needs a space before the colon.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -101,20 +102,36 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
-        if (primaryShardRouting.unassigned()) {
+
+        ShardRouting primaryShardRouting = null;
+        for (RoutingNode node: allocation.routingNodes()) {
```
Nit: space before colon

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -101,20 +102,36 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
+
+        ShardRouting primaryShardRouting = null;
+        for (RoutingNode node: allocation.routingNodes()) {
+            for (ShardRouting shard: node) {
```
Nit: space before colon

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -101,20 +102,36 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
-            allocation.routingTable().shardRoutingTable(index, shardId).replicaShardsWithState(ShardRoutingState.UNASSIGNED);
+        List<ShardRouting> replicaShardRoutings = new ArrayList<>();
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        for (ShardRouting shard: unassigned) {
```
Nit: space before colon

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java:
```diff
@@ -108,13 +108,22 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
+
+        ShardRouting shardRouting = null;
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        for (ShardRouting shard: unassigned) {
```
Nit: space before colon

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -101,20 +102,36 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
+        List<ShardRouting> replicaShardRoutings = new ArrayList<>();
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        for (ShardRouting shard: unassigned) {
+            if (shard.getIndexName().equals(index) && shard.getId() == shardId && !shard.primary()) {
```
The house style is to use `== false` instead of `!`.

DaveCTurner reviewed (dismissed):
Thanks @vigyasharma. This needs tests to support the change too: you need to construct some sequences of commands that fail because each command is ignoring the effects of the previous ones.

vigyasharma committed:
Addressed PR comments for styling

vigyasharma commented:
Thanks for the review. Made styling changes, will add test as well

vigyasharma committed:
Test for multiple conflicting allocation commands in single request

Verifies subsequent commands do not ignore the effects of previous ones.

vigyasharma commented:
@DaveCTurner Added test for multiple commands in single request. Please review when you get some time. Thanks!

DaveCTurner commented:
Thanks @vigyasharma. It's likely I won't be available to look at this in detail next week, but it's on my list for after that.

@elasticmachine ok to test

vigyasharma committed:
Remove unused import

vigyasharma commented:
@DaveCTurner Did you get a chance to review this?

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java:
```diff
@@ -110,13 +110,21 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
-        if (shardRouting.unassigned() == false) {
+
+        ShardRouting shardRouting = null;
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
```
Nit: you can inline `unassigned`, it's only used in the following line.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java:
```diff
@@ -101,20 +101,35 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
-            allocation.routingTable().shardRoutingTable(index, shardId).replicaShardsWithState(ShardRoutingState.UNASSIGNED);
+        List<ShardRouting> replicaShardRoutings = new ArrayList<>();
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
+        for (ShardRouting shard : unassigned) {
```
Nit: you can inline `unassigned`, it's only used in the following line.

DaveCTurner commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java:
```diff
@@ -108,13 +108,21 @@ public RerouteExplanation execute(RoutingAllocation allocation, boolean explain)
-        if (shardRouting.unassigned() == false) {
+
+        ShardRouting shardRouting = null;
+        RoutingNodes.UnassignedShards unassigned = allocation.routingNodes().unassigned();
```
Nit: you can inline `unassigned`, it's only used in the following line.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -677,4 +677,80 @@ public void testMoveShardFromNonDataNode() {
+                    new AllocateStalePrimaryAllocationCommand(index1, 0, node2, true)
+                ), false, false).getClusterState();
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), containsString("primary [" + index1 + "][0] is already assigned"));
```
This doesn't assert that the exception is actually thrown: the test still passes if the reroute succeeds. Suggest using `expectThrows()` like this:

```
        {
            final ClusterState finalClusterState = clusterState;
            assertThat(expectThrows(IllegalArgumentException.class, () -> {
                allocation.reroute(finalClusterState,
                    new AllocationCommands(
                        new AllocateStalePrimaryAllocationCommand(index1, 0, node1, true),
                        new AllocateStalePrimaryAllocationCommand(index1, 0, node2, true)
                    ), false, false);
            }).getMessage(), containsString("primary [" + index1 + "][0] is already assigned"));
        }
```

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -677,4 +677,80 @@ public void testMoveShardFromNonDataNode() {
+                new AllocationCommands(
+                    new AllocateStalePrimaryAllocationCommand(index1, 0, node1, true),
+                    new AllocateStalePrimaryAllocationCommand(index1, 0, node2, true)
+                ), false, false).getClusterState();
```
I think there's no need to call `getClusterState()` here.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -677,4 +677,80 @@ public void testMoveShardFromNonDataNode() {
+                new AllocationCommands(
+                    new AllocateEmptyPrimaryAllocationCommand(index2, 0, node1, true),
+                    new AllocateEmptyPrimaryAllocationCommand(index2, 0, node2, true)
+                ), false, false).getClusterState();
```
As above, no need for `.getClusterState()`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -677,4 +677,80 @@ public void testMoveShardFromNonDataNode() {
+                new AllocationCommands(
+                    new AllocateReplicaAllocationCommand(index3, 0, node2),
+                    new AllocateReplicaAllocationCommand(index3, 0, node2)),
+                false, false).getClusterState();
```
As above, no need for `.getClusterState()`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -677,4 +677,80 @@ public void testMoveShardFromNonDataNode() {
+                    new AllocateEmptyPrimaryAllocationCommand(index2, 0, node2, true)
+                ), false, false).getClusterState();
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(), containsString("primary [" + index2 + "][0] is already assigned"));
```
This should also use `expectThrows()`.

DaveCTurner commented on server/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java:
```diff
@@ -677,4 +677,80 @@ public void testMoveShardFromNonDataNode() {
+                    new AllocateReplicaAllocationCommand(index3, 0, node2)),
+                false, false).getClusterState();
+        } catch (IllegalArgumentException e) {
+            assertThat(e.getMessage(),
```
This should also use `expectThrows()`.

DaveCTurner reviewed (changes requested):
Thanks @vigyasharma this looks close. I requested a few tiny changes.

vigyasharma committed:
Fix test to verify expected exceptions

Test should assert that all expected exceptions are raised.
Includes styling changes from review.

vigyasharma commented:
Thanks @DaveCTurner. Made suggested changes.

DaveCTurner reviewed (approved):
LGTM thanks @vigyasharma. I expect to merge this after the weekend now.

DaveCTurner committed:
Merge branch 'master' into 2019-05-24-42066-fix-allocation-cmd

DaveCTurner committed:
Merge branch 'master' into fix-allocation-cmd
